### ISSUE TITLE
Use a single unique timestamp across scala versions per snapshot publish

### ISSUE THREAD

**Issue by mdedetrich** [apache/incubator-pekko#107]
As discussed in https://github.com/apache/incubator-pekko/issues/104#issuecomment-1384503408, due to a current limitation in sbt, when doing a cross publish amongst Scala versions using the `+` prefix it generates a separate timestamp for each Scala version where as ideally we should have a unique timestamp for an entire single publish operation.

A workaround as described in https://stackoverflow.com/a/21062454/1519631 is to pre-save the unique timestamp to a file and then read that timestamp along with a custom command alias.

**Comment by mdedetrich** [comment]
So since this is blocking a few things from pekko-http's side I have labelled this as a bug

**Comment by jrudolph** [comment]
The timestamp shouldn't be part of the artifact version. It is only added when the workspace is not clean. Otherwise, it will only be the commit id.

**Comment by mdedetrich** [comment]
@jrudolph This discussion is of relevance https://github.com/apache/incubator-pekko/pull/105#discussion_r1071912938 to your point. Currently @seglo set up nightlies which do contain a timestamp (which for a nightly makes sense) but I also want to setup snapshots which as you point out wouldn't need timestamp id's, but we need a strategy here because there is massive overlap between nightlies and snapshots (in my view nightlies are snapshots, just a specific way of handling them).

**Comment by jrudolph** [comment]
Indeed, I think the timestamp is currently included in the nightlies because the current main is not clean and while building the git workspace is changed, see

https://github.com/apache/incubator-pekko/actions/runs/3982251774/jobs/6826550691#step:4:173

and following.

Regarding, snapshots vs. nightlies, I agree. There's two potential solutions:

 * Either, push snapshots from nightlies, make sure to run only once per version on `main`
 * or, push snapshots for every commit and use nightlies only for more extensive testing than would be possible for each commit

I think we used the second option at least in akka-http.

**Comment by jrudolph** [comment]
> because the current main is not clean

The reason is probably that PR validation is not run for every Scala version (or at least `headerCheck` / `scalafmtCheckAll`).

**Comment by jrudolph** [comment]
```
pekko > +headerCheck
[info] Setting Scala version to 2.12.16 on 40 projects.
[info] Excluded 2 projects, run ++ 2.12.16 -v for more details.
[info] Reapplying settings...
[...]
[error] There are files without headers!
[error]   /home/johannes/git/opensource/pekko/actor-typed/src/main/scala-2.12/org/apache/pekko/actor/typed/internal/receptionist/Platform.scala
[error] There are files without headers!
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ByteString.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ByteIterator.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/compat/Future.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/package.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/dispatch/internal/ScalaBatchable.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/dispatch/internal/SameThreadExecutionContext.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/CompatImpl.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/compat/PartialFunction.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/ccompatUsedUntil213.scala
[error] There are files without headers!
[error]   /home/johannes/git/opensource/pekko/cluster-typed/src/main/scala-2.12/org/apache/pekko/cluster/typed/internal/receptionist/ClusterReceptionistProtocol.scala
[error] (actor-typed / Compile / headerCheck) There are files without headers!
[error]   /home/johannes/git/opensource/pekko/actor-typed/src/main/scala-2.12/org/apache/pekko/actor/typed/internal/receptionist/Platform.scala
[error] (actor / Compile / headerCheck) There are files without headers!
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ByteString.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ByteIterator.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/compat/Future.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/package.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/dispatch/internal/ScalaBatchable.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/dispatch/internal/SameThreadExecutionContext.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/CompatImpl.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/compat/PartialFunction.scala
[error]   /home/johannes/git/opensource/pekko/actor/src/main/scala-2.12/org/apache/pekko/util/ccompat/ccompatUsedUntil213.scala
[error] (cluster-typed / Compile / headerCheck) There are files without headers!
[error]   /home/johannes/git/opensource/pekko/cluster-typed/src/main/scala-2.12/org/apache/pekko/cluster/typed/internal/receptionist/ClusterReceptionistProtocol.scala
[error] Total time: 0 s, completed Jan 23, 2023 1:44:47 PM
pekko >
```

**Comment by mdedetrich** [comment]
Just to clarigy regarding `scamafmtCheckAll`, this has been removed because instead are using a github actions that runs a native GraalVM binary version of scalafmt, the reason to do this is because of developer ergonomics and speed (the idea with the github action is that it will be a mandated check before merging a PR).

**Comment by jrudolph** [comment]
> the idea with the github action is that it will be a mandated check before merging a PR

Yes, it should be a mandatory check but it also needs to be able to pick up all the relevant paths from sbt and don't think that will work right now with a project as complex as pekko. If we really care for speed in that particular case, we can cache the sbt cache of scalafmt in the Action, so it does not rerun on file versions seen before.

**Comment by mdedetrich** [comment]
> Yes, it should be a mandatory check but it also needs to be able to pick up all the relevant paths from sbt and don't think that will work right now with a project as complex as pekko.

It does actually pick up everything relevance, I actually checked. The part of the code that checks which paths are being formatted are part of core scalafmt which is defined in `.scalafmt.conf` and is common between scalafmt, sbt-scalafmt and scalafmt-native (this in fact was a deliberate decision by the devs of the latest 3.x version of scalafmt)

> If we really care for speed in that particular case, we can cache the sbt cache of scalafmt in the Action, so it does not rerun on file versions seen before.

At least for me, caching wouldn't solved the speed issue that I was experiencing the most. Its having to run scalafmt every time on small incremental changes. The current scalafmt github action typically takes 5-10 seconds to complete on the CI and also runs concurrently from the rest of the build.

I think this is going a bit offtopic though, feel free to make a thread on the dev mailing list where we can discuss this specifically

**Comment by jrudolph** [comment]
> That is it doesn't matter how scalafmt is run, whether its via sbt or on the cli with the JVM scalafmt runner or a GraalVM scalafmt, the results will be deterministic (and this to me is a feature, not a bug).

:+1: I see, we are using `--mode diff` which should be good enough for PR validation (as long as you cannot slip files past the PR check). Sorry for the noise.

I created #125 to run `headerCheck` on all Scala versions.

**Comment by mdedetrich** [comment]
> an untidy but workable solution to unblock work on pekko-http, etc. - https://github.com/pjfanning/micrometer-akka/pull/3/files#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R34-R43

So the problem here is that pekk-http already happens to have a custom built code for handling snapshots and at least currently I am unsure if I want to rip it all out/modify it, you can see it at https://github.com/apache/incubator-pekko-http/blob/main/project/AkkaDependency.scala, I wanted to ask @jrudolph about this to get the context of whats going on.

**Comment by jrudolph** [comment]
The last nightly snapshot was now generated without the timestamp (after
fixing the headers).

Matthew de Detrich ***@***.***> schrieb am Mo., 23. Jan.
2023, 17:27:

**Comment by mdedetrich** [comment]
@pjfanning In practice I would say this isn't really a blocker for the v1.0.0-M0 release because this behaviour only occurs when a version is derived from a state when `pullRequestValidation` fails which has now been solved thanks to @jrudolph .

For this reason its not really blocking anything as of now, I would only see this as being useful when we end up implementing the functionality to create snapshots from an ongoing unmerged PR but I would consider this as a nice to have to implement sometime in the future.

Do you agree to remove this from the milestone?

**Comment by pjfanning** [comment]
@mdedetrich is this closeable?

**Comment by mdedetrich** [comment]
Yes
